### PR TITLE
Add pwm audio sound support to bcm2837

### PR DESCRIPTION
**Pull request by gaotianyu1350**
Add pwm_sound_device, add some consts about pwm, dma

DMA works, but still cannot make sound.

suggest: maybe open a dev branch?

### CODE REVIEW

File: src/consts.rs [hunk 1]
```diff
@@ -5,3 +5,117 @@ pub const KERNEL_OFFSET: usize = 0xFFFF_0000_0000_0000;
 
 pub const RAW_IO_BASE: usize = 0x3F00_0000;
 pub const IO_BASE: usize = KERNEL_OFFSET + RAW_IO_BASE;
+
+
+pub const CLOCK_FREQ	: usize	= 500000000;		// PLLD
+pub const CLOCK_DIVIDER	: usize	= 2;
+
+pub const ARM_IO_BASE	: usize = KERNEL_OFFSET + 0x3F000000;
```
Comment (equation314) [inline src/consts.rs]: `ARM_IO_BASE` can be replaced with `IO_BASE `.

File: src/consts.rs [hunk 2]
```diff
@@ -5,3 +5,117 @@ pub const KERNEL_OFFSET: usize = 0xFFFF_0000_0000_0000;
...
+pub const ARM_DMA_BASE : usize = ARM_IO_BASE + 0x7000;
+// pub const ARM_DMA_CHAN_BASE : usize = KERNEL_OFFSET + 0x7E007000;  
+pub const GPU_IO_BASE	: usize = KERNEL_OFFSET + 0x7E000000;
```
Comment (equation314) [inline src/consts.rs]: I think the address starts from `0x7E000000` is same as `0x3F000000`, so could the constant `GPU_IO_BASE ` be removed?

File: src/consts.rs [hunk 3]
```diff
@@ -5,3 +5,117 @@ pub const KERNEL_OFFSET: usize = 0xFFFF_0000_0000_0000;
...
+
+pub const GPU_CACHED_BASE	: usize = KERNEL_OFFSET + 0x40000000;
+pub const GPU_UNCACHED_BASE	: usize = KERNEL_OFFSET + 0xC0000000;
+pub const GPU_MEM_BASE : usize = GPU_UNCACHED_BASE;
+
+//
+// PWM control register
```
Comment (equation314) [inline src/consts.rs]: Move these constants to the corresponding peripherals fils, and indicate the referenced documentation page.

File: src/pwm_sound_device.rs [hunk 4]
```diff
@@ -0,0 +1,494 @@
+use crate::consts::*;
```
Comment (equation314) [inline src/pwm_sound_device.rs]: Can this file be split into `pwm.rs` and `dma.rs` as peripherals implementation, and implement the sound device in rCore kernel?

File: src/pwm_sound_device.rs [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+// macro_rules! ARM_DMACHAN_DEBUG { ($( $chan:expr ),*)	=>	(ARM_DMA_BASE + (($chan) * 0x100) + 0x20) }
+
+fn CleanAndInvalidateDataCacheRange(nAddress: usize, nLength: usize) {
+
+}
+
+fn ARM_DMACHAN_CONBLK_AD(chan: usize) -> usize  {
+    ARM_DMA_BASE + (chan * 0x100) + 0x04
+}
+fn BUS_ADDRESS(addr: usize) -> usize {
```
Comment (equation314) [inline src/pwm_sound_device.rs]: This utility function can be moved into `const.rs` and have a more appropriate name.

File: src/pwm_sound_device.rs [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+            (*self.m_pControlBlock[1]).nNextControlBlockAddress = BUS_ADDRESS(self.m_pControlBlock[0] as usize) as u32;
+
+            // start clock and PWM device
+            self.RunPWM();
+
+            // enable and reset DMA channel
+            // PeripheralEntry();
+
+            // assert! (self.m_nDMAChannel <= DMA_CHANNEL_MAX);
+            write32(ARM_DMA_ENABLE, read32(ARM_DMA_ENABLE) | (1 << self.m_nDMAChannel));
```
Comment (equation314) [inline src/pwm_sound_device.rs]: Use `struct registers` shown in other peripherals to read and write IO memory.

File: src/util.rs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+pub unsafe fn read64(nAddress: * const usize) -> usize {
+    return *nAddress;
+}
+
+pub unsafe fn write64(nAddress: * mut usize, nValue:usize) {
+    *nAddress = nValue;
+}
+*/
+
+pub unsafe fn read32(nAddress: usize) -> u32 {
```
Comment (equation314) [inline src/util.rs]: Avoid direct unsafe reading and writing memory, replaced with `struct Registers`.

**Review by equation314 — COMMENTED** [review]
`cargo fmt` should be run. Warnings should be eliminated.

**Comment by jiegec** [comment]
Updates requested for @gaotianyu1350 .

**Comment by gaotianyu1350** [comment]
I have modified the code and please review the PR request. @equation314

**Pull request merged**